fanctrl.py

```diff
@@ -1,146 +1,215 @@
+                            client_socket.sendall(self.getCurrentStrategy().name.encode())
+                        except InvalidStrategyException:
+                            client_socket.sendall(("Error: unknown strategy: " + strategy).encode())
+                    if args.pause:
+                        self.pause()
+                        client_socket.sendall("Success".encode())
+                    if args.resume:
+                        self.resume()
+                        client_socket.sendall("Success".encode())
+                    if args.query:
+                        client_socket.sendall(self.getCurrentStrategy().name.encode())
+                    if args.list_strategies:
+                        client_socket.sendall('\n'.join(self.configuration.getStrategies()).encode())
+                    if args.reload:
+                        self.configuration.reload()
+                        if self.overwrittenStrategy is not None:
+                            self.overwriteStrategy(self.overwrittenStrategy.name)
+                        client_socket.sendall("Success".encode())
+                except:
+                    pass
+                finally:
+                    client_socket.shutdown(socket.SHUT_WR)
+                    client_socket.close()
+        finally:
+            server_socket.close()
 
     def setSpeed(self, speed):
         self.speed = speed
         bashCommand = f"ectool fanduty {speed}"
         subprocess.run(bashCommand, stdout=subprocess.PIPE, shell=True)
 
-    def adaptSpeed(self):
-        currentTemp = self.temps[self._tempIndex]
-        currentTemp = min(
-            currentTemp, self.getMovingAverageTemperature(self.movingAverageInterval)
-        )
-        minPoint = self.speedCurve[0]
-        maxPoint = self.speedCurve[-1]
-        for e in self.speedCurve:
+    def getActualTemperature(self):
+        bashCommand = "ectool temps all"
+        rawOut = subprocess.run(bashCommand, stdout=subprocess.PIPE, shell=True, text=True).stdout
+        rawTemps = re.findall(r'\(= (\d+) C\)', rawOut)
+        higherTemps = sorted([x for x in [int(x) for x in rawTemps] if x > 0], reverse=True)[:3]
+
+        # safety fallback to avoid damaging hardware
+        if len(higherTemps) == 0:
+            return 50
+        return round(sum(higherTemps) / len(higherTemps), 1)
+
+    # return mean temperature over a given time interval (in seconds)
+    def getMovingAverageTemperature(self, timeInterval):
+        slicedTempHistory = [x for x in self.tempHistory if x > 0][-timeInterval:]
+        if len(slicedTempHistory) == 0:
+            return self.getActualTemperature()
+        return round(sum(slicedTempHistory) / len(slicedTempHistory), 1)
+
+    def getEffectiveTemperature(self, currentTemp, timeIntervam):
+        # the moving average temperature count for 2/3 of the effective temperature
+        return round((self.getMovingAverageTemperature(timeIntervam) * 2 + currentTemp) / 3, 1)
```

Comment:
Again, can you explain the reason for changing the previous algorithm  ?
Using  `min(currentTemp, movingAverageTemperature )` was allowing for a quick reaction once a long workload was over, while also resisting to react too fast for a short burst in activity (a few seconds of CPU being at 100% would not trigger a big jump in fan speed, which is nice)

Reply:
I understand, I tried to make it a bit more linear by using the curent temperature in 1/3 of the effective one as it seems to make the fan change in a more progressive way.

Would you like me to go back to the previous one or should we mix them to get the lower between actual temp and this algorithm?

Reply:
Thinking about it a bit more, I guess your way is actually ok too.
I'm going to try it a little bit first, but it should be fine like that

Reply:
no problem, take your time
your way of taking the lowest to ensure fans going down as soon as possible is great too, I might mix the two as suggested above if you like it this way

Reply:
After testing it a bit, I have to say I prefer the old way. I like that I know instantly when a task is done just from hearing the fan go down. With the new way, the fan slowdown lags a bit. Can you please combine both ?

Reply:
yes, sure I will do that